File: users-service/src/main/java/mr/cell/routesticker/usersservice/RestResponseEntityExceptionHandler.java
package mr.cell.routesticker.usersservice;

import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import javax.servlet.http.HttpServletRequest;

@ControllerAdvice
public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler({ResourceNotFoundException.class})
    public ResponseEntity<Object> handleResourceNotFound(final ResourceNotFoundException ex,
                                                         final HttpServletRequest request,
                                                         final WebRequest webRequest) {
        final ErrorMessage errorMessage = new ErrorMessage(
                "Resource not found. Resource name: '" + ex.getResourceClassSimpleName() +
                        "', resource id: '" + ex.getResourceId() + "'",
                System.currentTimeMillis());
        final HttpHeaders httpHeaders = createBaseHttpHeaders();
        return handleExceptionInternal(ex, errorMessage, httpHeaders, HttpStatus.NOT_FOUND, webRequest);
    }

    @ExceptionHandler({IllegalArgumentException.class})
    public ResponseEntity<Object> handleIllegalArgumentException(final IllegalArgumentException ex,
                                                                 final HttpServletRequest request,
                                                                 final WebRequest webRequest) {
        final ErrorMessage errorMessage = new ErrorMessage("", System.currentTimeMillis());
        final HttpHeaders httpHeaders = createBaseHttpHeaders();
        return handleExceptionInternal(ex, errorMessage, httpHeaders, HttpStatus.BAD_REQUEST, webRequest);
    }

    private HttpHeaders createBaseHttpHeaders() {
        final HttpHeaders httpHeaders = new HttpHeaders();
        httpHeaders.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_UTF8_VALUE);
        return httpHeaders;
    }
}

File: climbing-routes-service/src/main/kotlin/mr/cell/routesticker/climbingroutesservice/RestResponseEntityExceptionHandler.kt
package mr.cell.routesticker.climbingroutesservice

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.server.ServerWebExchange

@ControllerAdvice
class RestResponseEntityExceptionHandler {

    @ExceptionHandler
    fun handleIllegalArgumentException(ex: IllegalArgumentException, exchange: ServerWebExchange):
            ResponseEntity<ErrorMessage> {
        val path = exchange.request.path.value()
        val errorMessage = ErrorMessage(System.currentTimeMillis(), ex.message ?: "", path)
        return ResponseEntity(errorMessage, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler
    fun handleResourceNotFoundException(ex: ResourceNotFoundException, exchange: ServerWebExchange):
            ResponseEntity<ErrorMessage> {
        val path = exchange.request.path.value()
        val message = "Resource not found. Resource name: '${ex.resourceClassSimpleName}', " +
                "resource id: '${ex.resourceId}'"
        val errorMessage = ErrorMessage(System.currentTimeMillis(), message, path)
        return ResponseEntity(errorMessage, HttpStatus.NOT_FOUND)
    }
}